### PR TITLE
Rule tracer

### PR DESCRIPTION
**Pull request by nanoMichal**
Rule Tracer, replaces rule debugger rule. (https://github.com/kgrgreer/foam3/pull/950) 
tracer objects can be added to the context, tracer logic is executed during rule evaluation at specific points in execution.

### CODE REVIEW

File: src/foam/nanos/ruler/RuleEngine.java [hunk 1]
```diff
@@ -60,23 +62,31 @@ public void execute(List<Rule> rules, FObject obj, FObject oldObj) {
     CompoundContextAgency compoundAgency = new CompoundContextAgency();
     ContextualizingAgency agency         = new ContextualizingAgency(compoundAgency, userX_, getX());
     Logger                logger         = (Logger) getX().get("logger");
+    RuleTracer            tracer         = (RuleTracer) ((FObject) getX().get("ruleTracer")).fclone();
```
Comment (nanoscott) [inline src/foam/nanos/ruler/RuleEngine.java]: you shouldn't clone something that can be null

File: src/foam/nanos/ruler/RuleEngine.java [hunk 2]
```diff
@@ -60,23 +62,31 @@ public void execute(List<Rule> rules, FObject obj, FObject oldObj) {
...
+    if ( tracer == null ) {
+      tracer = new NullTracer();
+      logger.debug("WARNING! No Rule Tracer Specified, using Null tracer");
```
Comment (nanoscott) [inline src/foam/nanos/ruler/RuleEngine.java]: Isn't this the default case .. not sure a debug message is needed. Considering the number of times the RuleEngine is hit for most puts, this message is going to be very frequent.

File: src/foam/nanos/ruler/RuleEngine.java [hunk 3]
```diff
@@ -176,45 +194,51 @@ public Object getResult(String ruleId) {
...
-
+    boolean permission = true;
     if ( user != null ) {
       var auth = (AuthService) getX().get("auth");
-      return auth.checkUser(getX(), user, "rule.read." + rule.getId());
+      permission = auth.checkUser(getX(), user, "rule.read." + rule.getId());
     }
-
-    return true;
+    tracer.tracePermission(permission);
```
Comment (nanoscott) [inline src/foam/nanos/ruler/RuleEngine.java]: can you trace the user as well?

File: src/foam/nanos/ruler/RuleEngine.java (hunk 2 above)
Comment (nanoMichal) [reply src/foam/nanos/ruler/RuleEngine.java]: fair point

File: src/foam/nanos/ruler/RuleEngine.java (hunk 3 above)
Comment (nanoMichal) [reply src/foam/nanos/ruler/RuleEngine.java]: Yes, added.

File: src/foam/nanos/ruler/RuleEngine.java [hunk 4]
```diff
@@ -60,23 +62,33 @@ public void execute(List<Rule> rules, FObject obj, FObject oldObj) {
     CompoundContextAgency compoundAgency = new CompoundContextAgency();
     ContextualizingAgency agency         = new ContextualizingAgency(compoundAgency, userX_, getX());
     Logger                logger         = (Logger) getX().get("logger");
+    RuleTracer            tracer         = (RuleTracer) getX().get("ruleTracer");
+    if ( tracer == null ) {
+      tracer = new NullTracer();
+    }
+    else {
+      tracer = (RuleTracer) ((FObject) tracer).fclone();
+    }
```
Comment (nanoscott) [inline src/foam/nanos/ruler/RuleEngine.java]:
```suggestion
    tracer = ( tracer == null ) ? new NullTracer() : (RuleTracer) ((FObject) tracer).fclone();
```

File: src/foam/nanos/ruler/RuleEngine.java [hunk 5]
```diff
@@ -115,11 +128,16 @@ public void execute(List<Rule> rules, FObject obj, FObject oldObj) {
   public void probe(List<Rule> rules, RulerProbe rulerProbe, FObject obj, FObject oldObj) {
     PM pm = PM.create(getX(), RulerProbe.getOwnClassInfo(), "Probe:" + obj.getClassInfo());
 
+    RuleTracer tracer = (RuleTracer) ((FObject) getX().get("ruleTracer")).fclone();
```
Comment (nanoscott) [inline src/foam/nanos/ruler/RuleEngine.java]: same null problem here.

**Review by nanoscott — COMMENTED** [review]
LGTM .. fancy!

File: src/foam/nanos/ruler/tracer/RuleTracer.js [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+  documentation: `
+    A model should implement this interface if it is to act as a Rule Tracer, meaning it is
+    called upon by the rule engine during certain operations.
+  `,
+
+  methods: [
+    {
+      name: 'preExecute',
+      documentation: `
+        This method happens when the rule permission is checked
```
Comment (lchanmann) [inline src/foam/nanos/ruler/tracer/RuleTracer.js]: same documentation "This method happens when the rule permission is checked" for several methods here?

File: src/foam/nanos/ruler/tracer/LoggingTracer.js [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+      args: [
+        { name: 'rule', type: 'foam.nanos.ruler.Rule' },
+        { name: 'obj', type: 'foam.core.FObject' },
+        { name: 'oldObj', type: 'foam.core.FObject' }
+      ],
+      javaCode: `
+        setLog(getLog().append("\\nRule: "));
+        setLog(getLog().append(rule.getName()));
+        setLog(getLog().append(", id: "));
+        setLog(getLog().append(rule.getId()));
```
Comment (lchanmann) [inline src/foam/nanos/ruler/tracer/LoggingTracer.js]: getLog().append("...") should be sufficient, might not need setLog(...)?

File: src/foam/nanos/ruler/RuleEngine.java [hunk 8]
```diff
@@ -60,23 +62,28 @@ public void execute(List<Rule> rules, FObject obj, FObject oldObj) {
     CompoundContextAgency compoundAgency = new CompoundContextAgency();
     ContextualizingAgency agency         = new ContextualizingAgency(compoundAgency, userX_, getX());
     Logger                logger         = (Logger) getX().get("logger");
+    RuleTracer            tracer         = (RuleTracer) getX().get("ruleTracer");
+    tracer = ( tracer == null ) ? new NullTracer() : (RuleTracer) ((FObject) tracer).fclone();
```
Comment (lchanmann) [inline src/foam/nanos/ruler/RuleEngine.java]: why do we need fclone()?

File: src/foam/nanos/ruler/RuleEngine.java [hunk 9]
```diff
@@ -115,11 +123,16 @@ public void execute(List<Rule> rules, FObject obj, FObject oldObj) {
   public void probe(List<Rule> rules, RulerProbe rulerProbe, FObject obj, FObject oldObj) {
     PM pm = PM.create(getX(), RulerProbe.getOwnClassInfo(), "Probe:" + obj.getClassInfo());
 
+    RuleTracer tracer = (RuleTracer) getX().get("ruleTracer");
+    tracer = ( tracer == null ) ? new NullTracer() : (RuleTracer) ((FObject) tracer).fclone();
```
Comment (lchanmann) [inline src/foam/nanos/ruler/RuleEngine.java]: why fclone?

File: src/foam/nanos/ruler/RuleEngine.java (hunk 8 above)
Comment (nanoMichal) [reply src/foam/nanos/ruler/RuleEngine.java]: We need to Fclone in the case of multiple rule engines being encountered in a dao put. Also a rule tracer can be set up with some starting property values, and we might want to use the same values for each. basically the instance of the tracer should be used once per run through the rule engine. Such as in a pmTracer for example.

File: src/foam/nanos/ruler/tracer/LoggingTracer.js [hunk 10]
```diff
@@ -55,10 +55,11 @@
...
       javaCode: `
-        setLog(getLog().append("\\nRule: "));
-        setLog(getLog().append(rule.getName()));
-        setLog(getLog().append(", id: "));
-        setLog(getLog().append(rule.getId()));
+        StringBuilder log = getLog();
+        log.append("\\nRule: "));
+        log.append(rule.getName()));
+        log.append(", id: "));
+        setLog(log.append(rule.getId()));
```
Comment (nanoscott) [inline src/foam/nanos/ruler/tracer/LoggingTracer.js]: Well if you're going to optimize this .. might as well do it right:
```suggestion
        setLog(getLog()
         .append("\\nRule: "))
          .append(rule.getName()))
          .append(", id: "))
          .append(rule.getId()));
```

File: src/foam/nanos/ruler/RuleEngine.java (hunk 8 above)
Comment (kgrgreer) [reply src/foam/nanos/ruler/RuleEngine.java]: Can you make NullTracer() be a singleton so we can just do NullTracer.instance() and avoid creating new ones all of the time?

**Pull request closed without merging**